### DISCUSSION TITLE
Shared roster are duplicated and falls to Not Found after migrating to v21.01

### DISCUSSION THREAD

**Discussion by ibaklykov** [processone/ejabberd#3606] [Questions] [answered]
We upgraded our ejabberd from 18.12 to 21.01 and encountered a problem.
When we make some changes in a shared roster and save it, the roster are duplicated.
Both in our mysql database and in the web GUI duplicate lines appear and if you click on any of them you will see Not Found message. Also this shared roster disappears from client apps (became invisible).
Please advice how to fix it. Any ideas would be helpful.

**Comment by badlop** [comment]
Can you update your post detailing:
* the options you have in ejabberd.yml related to mod_shared_roster, or any other that may affect it
* what exact ejabberd version you are using: how did you install it, from where, and if you added custom source code changes, modules, etc that may affect it
* exact reproduction steps: how can i reproduce exactly what you do to reproduce the problem? do you use webadmin or ejabberdctl? what exact arguments?

**Reply by ibaklykov** [reply]
Thank you for your reply.

- we don't have any options in mod_shared_roster block, it is empty:
`  mod_shared_roster: {}
`
- ejabberd version 21.01, installed from script ejabberd-21.01-linux-x64.run on clean debian 10
`./ejabberd-21.01-linux-x64.run
`
Nothing custom were added, just options to use mysql as database. Then I made a sql dump from our old ejabberd database and load the data to a new instance of ejabberd database. Maybe I had to convert my data and I missed some steps...
- I'm not sure you could reproduce this problem... I suppose the problem is in my old data and it would not show up on clean install. 
We use webadmin only, nothing else.

**Reply by prefiks** [reply]
Could you show duplicate lines from your database?

@badlop Possibly this is result of name -> label change that was done some time ago?

**Reply by ibaklykov** [reply]
Duplicate lines look like 

> | Marketing Dept                              | [{label,<<208,148,208,181,208,191,208,176,209,128,209,130,208,176,208,188,208,
          181,208,189,209,130,32,208,188,208,176,209,128,208,186,208,181,209,
          130,208,184,208,189,208,179,208,176>>},
 {displayed_groups,[<<208,144,208,180,208,188,208,184,208,189,208,184,209,129,
                      209,130,209,128,208,176,209,134,208,184,209,143>>,
...
                    <<208,174,209,128,208,184,208,180,208,184,209,135,208,
                      181,209,129,208,186,208,184,208,185,32,208,190,209,
                      130,208,180,208,181,208,187>>]}] | 2021-05-18 12:13:00 |

I'm not putting the second one - they are identical. By the way, I recreated this roster (in hope that it would help, but it did not). The other lines look different

> | IT Dept                                     | [{name,<<208,148,208,181,208,191,208,176,209,128,209,130,208,176,208,188,208,
         181,208,189,209,130,32,208,184,208,189,209,132,208,190,209,128,208,
         188,208,176,209,134,208,184,208,190,208,189,208,189,209,139,209,133,
         32,209,130,208,181,209,133,208,189,208,190,208,187,208,190,208,179,
         208,184,208,185>>},
 {displayed_groups,[<<208,144,208,180,208,188,208,184,208,189,208,184,209,129,
                      209,130,209,128,208,176,209,134,208,184,209,143>>,
...
                    <<208,174,209,128,208,184,208,180,208,184,209,135,208,
                      181,209,129,208,186,208,184,208,185,32,208,190,209,
                      130,208,180,208,181,208,187>>]}] | 2019-07-02 17:00:58 |

It's obviously connected with name -> label change, but how can I fix it?

**Reply by badlop** [reply]
I think I can reproduce that problem: using recent ejabberd, MySQL storage, clean database, WebAdmin to create new shared roster group, then in that new group options, click the "Submit" button... ejabberd crashes, shows an error message in the log file, webadmin shows "Not Found", and the MySQL table "sr_group" gets a duplicated group.

**Reply by prefiks** [reply]
Ha, i only was testing with srg_* commands, what error are you getting?

**Reply by badlop** [reply]
As it's easy to reproduce, I'm running "git bisect" to find the exact commit, will report here when find it. BTW, disabling cache doesn't help. Works correctly with mnesia.

**Reply by ibaklykov** [reply]
@badlop Yes, this is very similar to my case, but my ejabberd didn't crash.

**Reply by badlop** [reply]
Ah, sorry: when I said "ejabberd crashes" I wanted to say "ejabberd shows some error messages in the log file". Obviously it doesn't crash compltetely :)

PD: still compiling old ejabberd versions to check when the problem was introduced...

**Reply by badlop** [reply]
I'll continue investigating tomorrow.

ejabberd 20.07 seems to work correctly.
20.12 fails like this when clicking on "Submit" button on WebAdmin of a shared roster group options page:
```
2021-05-18 17:49:20.349863+02:00 [error]
Hook webadmin_page_host crashed when running mod_shared_roster:webadmin_page/3:
** exception error: bad argument
   in function  lists:keysearch/3
      called as lists:keysearch(label,1,error)
      *** argument 3: not a list
   in call from mod_shared_roster:get_opt/3 (/home/bernar/e/git/ejabberd/src/mod_shared_roster.erl, line 1153)
   in call from mod_shared_roster:shared_roster_group/4 (/home/bernar/e/git/ejabberd/src/mod_shared_roster.erl, line 979)
   in call from mod_shared_roster:webadmin_page/3 (/home/bernar/e/git/ejabberd/src/mod_shared_roster.erl, line 897)
   in call from ejabberd_hooks:safe_apply/4 (/home/bernar/e/git/ejabberd/src/ejabberd_hooks.erl, line 236)
   in call from ejabberd_hooks:run_fold1/4 (/home/bernar/e/git/ejabberd/src/ejabberd_hooks.erl, line 217)
   in call from ejabberd_web_admin:process_admin/3 (/home/bernar/e/git/ejabberd/src/ejabberd_web_admin.erl, line 551)
   in call from ejabberd_http:process/2 (/home/bernar/e/git/ejabberd/src/ejabberd_http.erl, line 373)
```

**Reply by ibaklykov** [reply]
ОК. Thank you. I'll be waiting for your conclusion.

**Reply by prefiks** [reply]
@badlop What group name you used? I can't reproduce that with master version. I can get this error when i change url to use unknown group name, but it works for me when using something else, maybe question of some normalization?

**Comment by prefiks** [comment, marked as answer]
Ok, i see what we get duplicate entries in sr_group, we don't have group name as primary key/unique index on it, so replace statement that we use to update doesn't work correctly here.

**Reply by prefiks** [reply]
I commited 95fa43aa96514b7e8b77fa7c29d2c0b5b1c1331a that adds those missing indexes. You can add them with
```
CREATE UNIQUE INDEX i_sr_group_name ON sr_group(name); 
```
or if you use new schema (when there is server_host column in sr_group table:)
```
CREATE UNIQUE INDEX i_sr_group_sh_name ON sr_group(server_host(191), name);
```

**Reply by prefiks** [reply]
And this did occur just now because, code that did upsert on mysql war changed from doing delete then insert to use replace statement, which needs primary keys to deteremine rows that need to be updated

**Reply by ibaklykov** [reply]
Thanks a lot! Adding an index (first one in my case) solved the problem.

**Reply by badlop** [reply]
Great! Your fix works for me, both updating the previous database, or creating a new one.

The commit that introduced this problem is  https://github.com/processone/ejabberd/commit/5b0f0d8352d729d953d84d2119989e3b93ca0139 ? In that case, it didn't affect any ejabberd release.

Anyway, the SQL update procedure will be mentioned in the next release announcement.
